designs/2019-expose-rules-to-formatters/readme.MD

```diff
@@ -0,0 +1,159 @@
+- Start Date: 2019-01-16
+- RFC PR: (leave this empty, to be filled in later)
+- Authors: Chris Meyer (@EasyRhinoMSFT)
+
+# Providing Rule Metadata to Formatters
+
+## Summary
+
+This proposal describes a design enhancement that provides formatters with details about the rules that have been executed by ESLint.
+
+## Motivation
+
+Currently, formatters only see the ID of each rule for which a violation was identified, plus an instance-specific description, as properties on each result object. Formatters are not able to access useful rule metadata, such as category, description, and help URL. Formatters are also not aware of the full set of rules that were run, information that may be useful in some cases.
+
+## Detailed Design
+
+Design Summary
+1. In `cli.js::printResults`, obtain the rules map from the `Engine` object.
+2. Add a second argument to the formatter's exported interface function. The value should be an object with a `rulesMetadata` property that is a map with the rule name as the key and the `rule.meta` object as the value. See the "Command Line Interface (cli.js) Changes" section below for implementation.
+
+We should use a container object as the argument, with a ruleId/rule.meta map as a property, in order to accommodate potential future expansions of the data we pass to formatters. This suggestion was previously made in the discussion of issue [#9841](https://github.com/eslint/eslint/issues/9841).
+
+### Command Line Interface (cli.js) Changes
+The implementation of this feature is very simple and straightfoward. The code location that invokes the formatter's exported interface function already has access to the API it should use to obtain the list of all executed rules. The call to `Engine.getRules` must be made in the try block because `engine` may be null during unit testing.
+
+```js
+function printResults(engine, results, format, outputFile) {
+    let formatter;
+    let rules;
+
+    try {
+        formatter = engine.getFormatter(format);
+        rules = engine.getRules();
+    } catch (e) {
+        log.error(e.message);
+        return false;
+    }
+
+    const rulesMetadata = {};
```

Comment:
I think we'd be better off using a `Map` here. It should (in theory) be more performant because we won't be changing the shape of an object by adding hundreds of keys to it.

Reply:
I think an object would be better because it would be more easily serializable by formatters like `json-with-metadata` (and anyone using `json-with-metadata` would be able to use almost the same interface as regular formatters).

I suspect there wouldn't be too much of a performance impact since the final shape of the object would be roughly the same on each iteration.

Reply:
I did a simple memory usage profile and it averages about 125 KB more with this change. I analyzed the three benchmark test scripts with the default ruleset.

Reply:
@EasyRhinoMSFT can you clarify that a bit? Are you saying `Map` used more memory or using an object did?

Reply:
@EasyRhinoMSFT this is the comment I referred to in my other comment. :)

I'd also like to hear a few other opinions or whether the rule meta should be in an object or a map. I'm of the mind that when we expose an object in our API that is intended to be used as a map, we should just be using a map. Most formatters will not be formatting the result as JSON, so I don't think optimizing for that case makes a lot of sense.

Reply:
I didn't actually do the Map case, just object vs. nothing. I'll profile the Map version today, though there was enough variance across trials that I doubt we'll see an appreciable difference.
The method was to look at process.memoryUsage() -- really basic and probably naive, so let me know if there is a different approach I should use. I ran trials with and without the change for each of the three bench scripts in .\tests\bench.

Reply:
Map is easier to insert data into, object is easier to deal with for consumers (serialization), so in my opinion, unless there's a rather large performance gain by using Map, I would go with an object.